### PR TITLE
reusing AST::Attribute and remove HIR::Attribute

### PR DESCRIPTION
**Pull request by thomasyonug**
Deleting HIR::Attribute and reusing AST::Attribute. 

Some relative reusing classes below:
1. DelimTokenTree
2. AttrInputMetaItemContainer
3. AttrInputLiteral
4. MetaItemLitExpr
5. MetaItemPathLit
6. MetaItemPath
7. MetaItemSeq
8. MetaWord
9. MetaNameValueStr
10. MetaListPath
11. MetaListNameValueStr
12. Attrbute
13. AttrInput
14. MacroParser

### CODE REVIEW

**Comment by philberty** [comment]
Nice work getting it all wired up! It might not have time to get around to reviewing this today but I will make sure to get back to you by Monday morning.

**Comment by philberty** [comment]
I've added other reviewers in the mean time might help get any feedback sooner on this change.

**Comment by thomasyonug** [comment]
Thanks @philberty . Take your time plz.

**Comment by YizhePKU** [comment]
Hmm, a while ago I made a similar PR(#256) about reusing operators in AST and HIR. We discussed it in issue #228. The decision then was to put operators in the global namespace because operator "does not belong more or less to the AST or the HIR, so a shared space would be best".

**Comment by YizhePKU** [comment]
Also this may be a good chance to introduce a cleanup: `using AttrVec = std::vector<AST::Attribute>`.

**Comment by thomasyonug** [comment]
> Hmm, a while ago I made a similar PR(#256) about reusing operators in AST and HIR. We discussed it in issue #228. The decision then was to put operators in the global namespace because operator "does not belong more or less to the AST or the HIR, so a shared space would be best".

Yeah, about the namespace problem. I mentioned that what if we move the common class like Attribute out of AST namespace. 
But Phil said that keeping Attribute within the AST namespace is still good, as it signifies it belongs to the AST from parsing the language. 
I think maybe we can discuss it again next week when he came back.

**Comment by YizhePKU** [comment]
> But Phil said that keeping Attribute within the AST namespace is still good, as it signifies it belongs to the AST from parsing the language.

That makes sense. It's not a big deal anyway.

File: gcc/rust/hir/tree/rust-hir-full-decls.h [hunk 1]
```diff
@@ -29,14 +29,9 @@ class TokenTree;
 class MacroMatch;
 class Token;
 struct Literal;
-class DelimTokenTree;
 class PathSegment;
 class SimplePathSegment;
 class SimplePath;
```
Comment (YizhePKU) [inline gcc/rust/hir/tree/rust-hir-full-decls.h:34]: Do you mean to remove `SimplePath` here?

File: gcc/rust/hir/tree/rust-hir-full-decls.h [hunk 2]
```diff
@@ -227,8 +220,6 @@ struct MacroTranscriber;
 struct MacroRule;
 class MacroRulesDefinition;
 class MacroInvocation;
-class MetaItemPath;
-class MetaItemSeq;
 class MetaWord;
```
Comment (YizhePKU) [inline gcc/rust/hir/tree/rust-hir-full-decls.h:223]: Do you mean to remove `Meta*` stuff here?

**Review by YizhePKU — CHANGES_REQUESTED** [review]

File: gcc/rust/hir/tree/rust-hir-full-decls.h (hunk 1 above)
Comment (YizhePKU) [reply gcc/rust/hir/tree/rust-hir-full-decls.h:34]: Everybody else now uses `AST::SimplePath`, correct?

Comment (thomasyonug) [reply gcc/rust/hir/tree/rust-hir-full-decls.h:34]: yeah, let me try to remove it.

Comment (thomasyonug) [reply gcc/rust/hir/tree/rust-hir-full-decls.h:34]: LOL, I found this is another miss deletion. I already remove the SimplePath definition but decl.

**Comment by YizhePKU** [comment]
Suggestion: Treat force-push only as a last-resort. Force-pushing removes old files, making it harder for others to follow the conversation.

**Comment by thomasyonug** [comment]
> Suggestion: Treat force-push only as a last resort. Force-pushing removes old files, making it harder for others to follow the conversation.

Sorry, not notice that.

**Comment by thomasyonug** [comment]
> Beware that the first comment is used by bors when creating its own commit. It may be better to keep it more descriptive and keep other remarks as comments bellow ?

Thanks. About the other remarks do you mean #Fixes and #Address？

**Comment by thomasyonug** [comment]
> Beware that the first comment is used by bors when creating its own commit. It may be better to keep it more descriptive and keep other remarks as comments bellow ?

Got it.

**Comment by thomasyonug** [comment]
> Beware that the first comment is used by bors when creating its own commit. It may be better to keep it more descriptive and keep other remarks as comments bellow ?

Thanks, got it.

File: gcc/rust/backend/rust-compile-base.h [hunk 3]
```diff
@@ -36,8 +36,8 @@ class HIRCompileBase : public HIR::HIRVisitor
   // virtual void visit(TokenTree& token_tree) {}
   // virtual void visit(MacroMatch& macro_match) {}
   virtual void visit (HIR::Token &tok) {}
-  virtual void visit (HIR::DelimTokenTree &delim_tok_tree) {}
-  virtual void visit (HIR::AttrInputMetaItemContainer &input) {}
+  // virtual void visit (HIR::DelimTokenTree &delim_tok_tree) {}
```
Comment (dkm) [inline gcc/rust/backend/rust-compile-base.h:39]: What about all these commented definitions ? Can't we simply remove them?

File: gcc/rust/typecheck/rust-hir-type-check-base.h [hunk 4]
```diff
@@ -163,12 +162,12 @@ class TypeCheckBase : public HIR::HIRVisitor
   virtual void visit (HIR::MacroMatcher &) override {}
   virtual void visit (HIR::MacroRulesDefinition &) override {}
   virtual void visit (HIR::MacroInvocation &) override {}
-  virtual void visit (HIR::MetaItemPath &) override {}
-  virtual void visit (HIR::MetaItemSeq &) override {}
-  virtual void visit (HIR::MetaWord &) override {}
-  virtual void visit (HIR::MetaNameValueStr &) override {}
-  virtual void visit (HIR::MetaListPaths &) override {}
-  virtual void visit (HIR::MetaListNameValueStr &) override {}
+  // virtual void visit (HIR::MetaItemPath &) override {}
```
Comment (dkm) [inline gcc/rust/typecheck/rust-hir-type-check-base.h:165]: Same question here ?

Comment (thomasyonug) [reply gcc/rust/typecheck/rust-hir-type-check-base.h:165]: I thought that's better for consistency to comment the visit like before. But yes, I can simply remove them all.

**Comment by YizhePKU** [comment]
What about putting `AttrVec` in the global namespace(`Rust` namespace actually)? It's nicer to write `AttrVec` instead of `AST::AttrVec`.

File: gcc/rust/typecheck/rust-hir-type-check-base.h (hunk 4 above)
Comment (dkm) [reply gcc/rust/typecheck/rust-hir-type-check-base.h]: Maybe there's a reason for these comments (@philberty ?) but I guess they are not really needed (the visitor pattern being widely used in gccrs, it's easy to find which methods can be overridden)

**Comment by thomasyonug** [comment]
> What about putting `AttrVec` in the global namespace(`Rust` namespace actually)? It's nicer to write `AttrVec` instead of `AST::AttrVec`.

It's kind of weird for me that  two type symbols are in different namespace. 

But I think it's better to do that if some other type alias also in global namespace.

**Review by philberty — APPROVED** [review]

**Comment by philberty** [comment]
I think this looks good but before we can merge you will need to squash these commits. I see this change as a single logical commit.

**Comment by philberty** [comment]
I do like having this behind the AST, since attributes are driven by the AST. It was slightly different for the PR @YizhePKU referenced where we had the logical operators enum defined in Lexer, AST, HIR, GCC. So in that case it was best to just have 1.

**Comment by thomasyonug** [comment]
> I think this looks good but before we can merge you will need to squash these commits. I see this change as a single logical commit.

Sure.:)

**Pull request merged**